File: algorithm_libraries/python/data_structure/union_find.py
from typing import List


class UnionFind:
    def __init__(self, n: int) -> None:
        self.n = n
        self.parents = [-1] * n

    def find(self, x: int) -> int:
        if self.parents[x] < 0:
            return x
        else:
            self.parents[x] = self.find(self.parents[x])
            return self.parents[x]

    def unite(self, x: int, y: int) -> None:
        x = self.find(x)
        y = self.find(y)

        if x == y:
            return

        if self.parents[x] > self.parents[y]:
            x, y = y, x

        self.parents[x] += self.parents[y]
        self.parents[y] = x

    def size(self, x: int) -> int:
        return -self.parents[self.find(x)]

    def same(self, x: int, y: int) -> bool:
        return self.find(x) == self.find(y)

    def members(self, x: int) -> List[int]:
        root = self.find(x)
        return [i for i in range(self.n) if self.find(i) == root]

    def roots(self) -> List[int]:
        return [i for i, x in enumerate(self.parents) if x < 0]

    def group_count(self) -> int:
        return len(self.roots())


if __name__ == "__main__":
    N, Q = map(int, input().split())
    uf = UnionFind(N + 1)
    for i in range(Q):
        T, U, V = map(int, input().split())
        if T == 0:
            uf.unite(U, V)
        else:
            if uf.same(U, V):
                print(1)
            else:
                print(0)

File: algorithm_libraries/cpp/data_structure/union_find.cpp
#include <bits/stdc++.h>
using namespace std;

struct UnionFind {
    vector<long long> data;

    UnionFind(long long sz){
        data.assign(sz, -1);
    }
    long long find(long long k){
        if(data[k] < 0) return (k);
        return (data[k] = find(data[k]));
    }
    bool unite(long long x, long long y){
        x = find(x), y = find(y);
        if(x == y) return (false);
        if(data[x] > data[y]) swap(x, y);
        data[x] += data[y];
        data[y] = x;
        return (true);
    }
    long long size(long long k) {
        return (-data[find(k)]);
    }
    bool same(long long x, long long y){
        return (find(x) == find(y));
    }
    long long group_count() {
        long long res = 0;
        long long size = data.size();
        for(long long i = 0; i < size; i++) {
            if(data[i] < 0) res++;
        }
        return (res);
    }
};

int main() {
    long long N, Q;
    cin >> N >> Q;
    UnionFind uf(N + 1);
    for(long long i = 0; i < Q; i++) {
        long long T, U, V;
        cin >> T >> U >> V;
        if(T == 0) {
            uf.unite(U, V);
        } else {
            if(uf.same(U, V)) cout << 1 << endl;
            else cout << 0 << endl;
        }
    }
    return (0);
}